package.json

```diff
@@ -22,6 +22,7 @@
   "scripts": {
     "prettier-check": "ts-node ./scripts/prettier-check.ts",
     "prettier-fix": "prettier --write specification/**/*.json"
+    "postinstall": "tsc"
```

Comment:
I think the idea is this will run `npx tsc` which is prerequisite for durnning `npx prettier-fix`: https://github.com/Azure/azure-sdk-tools/issues/6017#issuecomment-1517401487

Reply:
I understand tsc compiles any source it finds in the repo but that seems like it is a little overkill just to get prettier working. What I'm trying to understand is do we have other options? Also I'd like to better understand the difference between this prettier script vs what we do in the unified pipeline which is slightly different I believe. We should see if we can make these exactly the same.

Reply:
I agree that running `tsc` in `postinstall` seems like overkill just to get `prettier` working.  I will try to understand what we are trying to do here, why it's broken if you don't run `tsc`, and the best way to fix it.

Reply:
An alternative would be to run `tsc` as a pre-step before the prettier scripts:

```
"scripts": {
  "preprettier-check": "tsc",
  "prettier-check": "ts-node ./scripts/prettier-check.ts",
```

However, running `tsc` from `scripts` fails for me, regardless of whether it's run in `preprettier-check` or `postinstall`:

```
~/azure-rest-api-specs$ npm install

> postinstall
> tsc -p /home/mharder/azure-rest-api-specs

node_modules/@types/node/ts4.8/test.d.ts:612:34 - error TS1005: '?' expected.

612             : F extends abstract new (...args: any) => infer T
                                     ~~~
```

But running `tsc` directly from the command line works with no errors:

```
~/azure-rest-api-specs$ tsc --build --verbose
[9:49:06 PM] Projects in this build:
    * tsconfig.json

[9:49:06 PM] Project 'tsconfig.json' is up to date because newest input 'scripts/prettier-check.ts' is older than output 'scripts/utils.js'
```

I think my recommendation would just be to tell users to manually run `tsc` before the prettier scripts.

Reply:
The problem with `tsc` above, is the version of `typescript` in `package.json` and `package-lock.json` is too old to compile the types installed by package `@types/node`.  It works when I run `tsc` directly from the command line, because I have a newer version of `typescript` installed globally.  However, running `tsc` from the `scripts` section uses the version of `typescript` installed locally, which is too old (`3.9.10`).  You can repro this from the command line by running `npx tsc` instead of `tsc`.

So I think I would recommend the following:

1. Update the version of `typescript` in `package.json` and `package-lock.json` to the latest compatible version.
2. Add steps `preprettier-check` and `preprettier-fix` that call `tsc`.  This makes these scripts a little slower since they need to run `tsc` every time, but I like this better than running `tsc` in a global `postinstall` when it's only needed for `prettier`.

Reply:
@weshaggard , the difference in unified pipeline is the scripts under `\scripts` folder is compiled JS but in spec repo they are `*.ts` file. That's why `tsc` command needs to run to turn them into JS. I don't worry about the perf as only \scripts folder might contain *.ts files. `specification` folder shouldn't have *.ts files.

@mikeharder , whatever the typescript version in package.json needs to be fixed. I would prefer to wrap `tsc` in `postinstall` command as it would be needed for other scenarios rather than `prettier` only.  @konrad-jamrozik , can you help with typescript version upgrade?

Reply:
I am fine with adding `tsc` to `postinstall`, I agree it doesn't add much time to the `npm install` command.